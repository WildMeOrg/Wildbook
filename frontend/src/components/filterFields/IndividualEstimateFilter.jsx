import React from "react";
import { FormattedMessage } from "react-intl";
import Description from "../Form/Description";
import FormGroupText from "../Form/FormGroupText";

export default function IndividualEstimateFilter({ store }) {
  return (
    <div>
      <h4>
        <FormattedMessage id="FILTER_INDIVIDUAL_ESTIMATE" />
      </h4>
      <Description>
        <FormattedMessage id="FILTER_INDIVIDUAL_ESTIMATE_DESC" />
      </Description>

      <FormGroupText
        label="FILTER_BEST_ESTIMATE_INDIVIDUALS"
        noDesc={true}
        field={"occurrenceBestGroupSizeEstimate"}
        term={"match"}
        filterId={"occurrenceBestGroupSizeEstimate"}
        filterKey={"Best Estimate Individuals"}
        store={store}
      />
      <FormGroupText
        label="FILTER_MINIMUM_ESTIMATE_INDIVIDUALS"
        noDesc={true}
        field={"occurrenceMinGroupSizeEstimate"}
        term={"match"}
        filterId={"occurrenceMinGroupSizeEstimate"}
        filterKey={"Minimum Estimate Individuals"}
        store={store}
      />
      <FormGroupText
        label="FILTER_MAXIMUM_ESTIMATE_INDIVIDUALS"
        noDesc={true}
        field={"occurrenceMaxGroupSizeEstimate"}
        term={"match"}
        filterId={"occurrenceMaxGroupSizeEstimate"}
        filterKey={"Maximum Estimate Individuals"}
        store={store}
      />
    </div>
  );
}
