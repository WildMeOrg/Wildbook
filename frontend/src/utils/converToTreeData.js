const convertToTreeData = (locationData) => {
  if (!Array.isArray(locationData)) return [];
  return locationData.map((location) => ({
    title: location.id,
    value: location.id,
    children:
      location.locationID?.length > 0
        ? convertToTreeData(location.locationID)
        : [],
  }));
};

export default convertToTreeData;
