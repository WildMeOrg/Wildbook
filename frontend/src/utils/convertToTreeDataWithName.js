const convertToTreeDataWithName = (locationData) => {
  if (!Array.isArray(locationData)) return [];
  return locationData.map((location) => ({
    title: location.name,
    value: location.id,
    children:
      location.locationID?.length > 0
        ? convertToTreeDataWithName(location.locationID)
        : [],
  }));
};

export default convertToTreeDataWithName;
