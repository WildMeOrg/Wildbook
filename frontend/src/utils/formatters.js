import { format } from "date-fns";

export const formatDate = (input, fancy = false, fallback = "") => {
  const formatter = fancy ? "PP" : "yyyy-MM-dd HH:mm";
  try {
    const jsDate = typeof input === "string" ? new Date(input) : input;
    const formattedDate = format(jsDate, formatter);
    return formattedDate;
  } catch (_error) {
    // console.error(error);
    return fallback;
  }
};
