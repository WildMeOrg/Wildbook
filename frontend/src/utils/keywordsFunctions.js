import axios from "axios";

export async function postRestKeyword(payload, opts = {}) {
  const res = await axios.post(`/RestKeyword`, payload, {
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
    },
  });
  return res.data; 
}

export async function getAssetKeywords(mid) {
  const res = await axios.get(`/RestKeyword`, {
    params: { assetIds: String(mid) },   
    withCredentials: true,
    headers: { Accept: "application/json" },
  });
  const data = res.data || {};
  const byMid = (data.results && data.results[mid]) || {};
  return Object.entries(byMid).map(([id, name]) => ({ id, name }));
}
export function addExistingKeyword(mid, keywordId) {
  return postRestKeyword({ onMediaAssets: { assetIds: [mid], add: [keywordId] } });
}

export function addNewKeywordText(mid, text) {
  return postRestKeyword({ onMediaAssets: { assetIds: [mid], newAdd: [text] } });
}

export function removeKeyword(mid, keywordId) {
  return postRestKeyword({ onMediaAssets: { assetIds: [mid], remove: [keywordId] } });
}
