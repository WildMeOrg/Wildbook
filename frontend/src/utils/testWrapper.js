import React from "react";
import { IntlProvider } from "react-intl";
import messages from "../locale/en.json";

export const wrapper = ({ children }) => (
  <IntlProvider locale="en" messages={messages}>
    {children}
  </IntlProvider>
);
