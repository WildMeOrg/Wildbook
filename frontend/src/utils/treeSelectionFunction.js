export function findNodeByValue(treeData, value) {
  for (const node of treeData || []) {
    if (node?.value === value) return node;
    if (node?.children?.length) {
      const found = findNodeByValue(node.children, value);
      if (found) return found;
    }
  }
  return null;
}

export function getAllDescendantValues(node) {
  const res = [];
  (node?.children || []).forEach((child) => {
    if (child?.value != null) res.push(child.value);
    res.push(...getAllDescendantValues(child));
  });
  return res;
}

export function expandIds(locationOptions, values = []) {
  const set = new Set();
  (values || []).forEach((val) => {
    if (!val) return;
    set.add(val);
    const node = findNodeByValue(locationOptions, val); 
    if (node) {
      getAllDescendantValues(node).forEach((cid) => {
        if (cid != null) set.add(cid);
      });
    }
  });
  return Array.from(set);
}