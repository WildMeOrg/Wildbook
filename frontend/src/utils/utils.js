import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import axios from 'axios';
import { QueryClient, QueryClientProvider } from 'react-query';
import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router-dom';
import messages from '../locale/en.json';
import AuthContext from '../AuthProvider';

jest.mock('axios');

const renderWithProviders = (ui, isLoggedIn = true) => {
  const queryClient = new QueryClient();
  return render(
    <QueryClientProvider client={queryClient}>
      <IntlProvider locale="en" messages={messages}>
      <AuthContext.Provider value={{ isLoggedIn }}>
        <MemoryRouter>{ui}</MemoryRouter>
      </AuthContext.Provider>
      </IntlProvider>
    </QueryClientProvider>
  );
};

const renderWithRouter = (ui, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);
  return render(ui, { wrapper: BrowserRouter });
};

const fireInput = (placeholder, value) => {
  const input = screen.getByPlaceholderText(placeholder);
  fireEvent.change(input, { target: { value } });
  return input;
};

const clickButton = (buttonText) => {
  fireEvent.click(screen.getByRole('button', { name: buttonText }));
};

const waitForText = async (text) => {
  await waitFor(() => expect(screen.getByText(text)).toBeInTheDocument());
};

const mockAxiosSuccess = (mockedData) => {
  axios.get.mockResolvedValue({ data: mockedData });
};

const mockAxiosFailure = () => {
  axios.get.mockRejectedValue(new Error('Network error'));
};

const mockComponent = (testId) => {
  return () => {
    const React = require("react");
    const MockLandingImage = () =>
      React.createElement("div", { "data-testid": testId });
    MockLandingImage.displayName = testId;
    return MockLandingImage;
  }
}

export * from '@testing-library/react';
export {
  renderWithRouter as render,
  fireInput,
  clickButton,
  waitForText,
  mockAxiosSuccess,
  mockAxiosFailure,
  renderWithProviders,
  mockComponent,
};
