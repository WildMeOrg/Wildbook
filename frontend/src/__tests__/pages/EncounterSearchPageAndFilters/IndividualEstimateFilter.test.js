import React from "react";
import { screen } from "@testing-library/react";
import IndividualEstimateFilter from "../../../components/filterFields/IndividualEstimateFilter";
import { renderWithProviders } from "../../../utils/utils";

const mockStore = {
  formFilters: [],
  addFilter: jest.fn(),
  removeFilter: jest.fn(),
};

describe("IndividualEstimateFilter Component", () => {
  beforeEach(() => {
    mockStore.formFilters = [];
    jest.clearAllMocks();
  });

  const renderComponent = () => {
    renderWithProviders(<IndividualEstimateFilter store={mockStore} />);
  };

  test("renders main label and description correctly", () => {
    renderComponent();

    expect(
      screen.getByText(/FILTER_INDIVIDUAL_ESTIMATE_DESC/),
    ).toBeInTheDocument();
    expect(
      screen.getByText(/FILTER_BEST_ESTIMATE_INDIVIDUALS/),
    ).toBeInTheDocument();
  });

  test("renders all FormGroupText components correctly", () => {
    renderComponent();

    expect(
      screen.getByText(/FILTER_BEST_ESTIMATE_INDIVIDUALS/),
    ).toBeInTheDocument();
    expect(
      screen.getByText(/FILTER_MINIMUM_ESTIMATE_INDIVIDUALS/),
    ).toBeInTheDocument();
    expect(
      screen.getByText(/FILTER_MAXIMUM_ESTIMATE_INDIVIDUALS/),
    ).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const { asFragment } = renderWithProviders(
      <IndividualEstimateFilter store={mockStore} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("calls store methods on interaction with FormGroupText components", () => {
    renderComponent();

    expect(mockStore.addFilter).not.toHaveBeenCalled();
    expect(mockStore.removeFilter).not.toHaveBeenCalled();
  });

  test("handles store state correctly", () => {
    mockStore.formFilters = [
      {
        filterId: "occurrenceBestGroupSizeEstimate",
        query: { match: { occurrenceBestGroupSizeEstimate: "10" } },
      },
    ];

    renderComponent();

    expect(
      screen.getByText(/FILTER_BEST_ESTIMATE_INDIVIDUALS/),
    ).toBeInTheDocument();
  });
});
