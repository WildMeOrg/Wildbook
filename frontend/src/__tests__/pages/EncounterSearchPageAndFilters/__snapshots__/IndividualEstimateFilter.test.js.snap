// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndividualEstimateFilter Component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h4>
      FILTER_INDIVIDUAL_ESTIMATE
    </h4>
    <div
      style="font-size: 14px; font-weight: 400; line-height: 21px; color: rgb(175, 179, 183); margin-bottom: 1rem;"
    >
      FILTER_INDIVIDUAL_ESTIMATE_DESC
    </div>
    <div
      class="mt-2"
    >
      <label
        class="form-label"
      >
        FILTER_BEST_ESTIMATE_INDIVIDUALS
      </label>
      <input
        class="form-control"
        placeholder="Type here"
        type="text"
        value=""
      />
    </div>
    <div
      class="mt-2"
    >
      <label
        class="form-label"
      >
        FILTER_MINIMUM_ESTIMATE_INDIVIDUALS
      </label>
      <input
        class="form-control"
        placeholder="Type here"
        type="text"
        value=""
      />
    </div>
    <div
      class="mt-2"
    >
      <label
        class="form-label"
      >
        FILTER_MAXIMUM_ESTIMATE_INDIVIDUALS
      </label>
      <input
        class="form-control"
        placeholder="Type here"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
